File: app/modules/zen-link-transport/android/src/main/java/expo/modules/zenlinktransport/PinnedEndpointRegistry.kt
package expo.modules.zenlinktransport

import java.util.concurrent.ConcurrentHashMap

// Native consumers use live listener ownership, never a JS assertion that an
// arbitrary loopback socket is encrypted. Every accepted stream still pins TLS.
object PinnedEndpointRegistry {
  private data class Entry(val owner: Any, val pin: String)
  private val entries = ConcurrentHashMap<Int, Entry>()

  internal fun register(port: Int, pin: String, owner: Any) {
    entries[port] = Entry(owner, pin.lowercase())
  }

  internal fun remove(port: Int, owner: Any) {
    entries.computeIfPresent(port) { _, entry -> if (entry.owner === owner) null else entry }
  }

  fun contains(port: Int, pin: String): Boolean = entries[port]?.pin == pin.lowercase()
}

File: app/modules/zen-link-transport/ios/PinnedEndpointRegistry.swift
import Foundation

public enum PinnedEndpointRegistry {
  private static let lock = NSLock()
  private static var entries: [Int: (UUID, Data)] = [:]

  static func register(port: Int, pin: Data, owner: UUID) {
    lock.lock(); defer { lock.unlock() }
    entries[port] = (owner, pin)
  }

  static func remove(port: Int, owner: UUID) {
    lock.lock(); defer { lock.unlock() }
    if entries[port]?.0 == owner { entries.removeValue(forKey: port) }
  }

  public static func contains(port: Int, pin: String) -> Bool {
    lock.lock(); defer { lock.unlock() }
    guard let entry = entries[port] else { return false }
    return entry.1.map { String(format: "%02x", $0) }.joined() == pin.lowercased()
  }
}
